Src/IronPython/Runtime/PythonContext.cs

```diff
@@ -936,125 +936,79 @@ private sealed class OptimizedBuiltinHasher {
             ContractUtils.Requires(stream.CanSeek && stream.CanRead, nameof(stream), "The stream must support seeking and reading");
 
             stream.Seek(0, SeekOrigin.Begin);
-            byte[] bomBuffer = new byte[3];
-            int bomRead = stream.Read(bomBuffer, 0, 3);
-            int bytesRead = 0;
-            bool isUtf8 = false;
-            if (bomRead == 3 && (bomBuffer[0] == 0xef && bomBuffer[1] == 0xbb && bomBuffer[2] == 0xbf)) {
-                isUtf8 = true;
-                bytesRead = 3;
-            } else {
-                stream.Seek(0, SeekOrigin.Begin);
-            }
 
+            Encoding sourceEncoding = null;
             string encodingName = null;
             int linesRead = 0;
+            bool hasBom = false;
             // sr is used to read the magic comments (PEP-263)
-            // default system ASCII encoding will never throw exceptions, converting bad bytes to '?' instead
-            // which is sufficient to parse the magic comments
-            // also StreamReader will automatically switch to proper Unicode encoding (non-throwing) when BOM present
-            using (StreamReader sr = new StreamReader(stream, Encoding.ASCII, detectEncodingFromByteOrderMarks: true, bufferSize: 1024, leaveOpen: true)) {
+            // Default system ASCII encoding will not throw exceptions, converting bad bytes to '?' instead
+            // which is sufficient to correctly parse the magic comments.
+            // Also StreamReader will automatically switch to proper Unicode encoding (non-throwing) when BOM present.
+            Encoding bootstrapEncoding = Encoding.ASCII;
+            using (StreamReader sr = new StreamReader(stream, bootstrapEncoding, detectEncodingFromByteOrderMarks: true, bufferSize: 1024, leaveOpen: true)) {
                 string line;
-                line = ReadOneLine(sr, ref bytesRead);
+                line = sr.ReadLine();
```

Comment:
CPython allows arbitrarily long lines in it's coding check?

Reply:
Never tested whether there is a limit on a comment line. But there is no limit on any other line, as `test_source_encoding.test_file_parse` attests.

Reply:
I haven't found it specified anywhere, but tests with CPython show that the lines are truncated after 511 characters. I guess we better follow this convention. This has to be done through a `StreamReader` though.

Reply:
This may result in a truncated encoding name, for instance `# ... coding: latin-15` becomes `latin-1` if `5` happens to be in the 512-th column.

Reply:
Correction: not 512 characters but 512 bytes. However, the BOM bytes do not count to the limit.

Reply:
What do you think of dropping support for widechar Unicode in source files (utf-16*, utf-32)? The code will become simpler (I won't have to use `StreamReader`). AFAIK, CPython does not support widechar Unicode.

Reply:
I would be fine with dropping wide character encodings since PEP263 seems to talk about requiring ASCII compatible encodings. Assuming it hasn't been added in later versions (e.g. 3.8) of Python that is.

We don't have to match the 512-byte limit exactly, I was just thinking that some sort of limit would be useful since we're supposed to be quickly probing the file for an encoding.

Reply:
Great. I could set the limit to 512 characters per line, which will be 512 bytes for non-unicode encodings, and 2K for UTF-32* at the high end (or an extreme and unlikely worst case for UTF-8).

However, I am thinking that if CPython ever adds support for widechar Unicode source code, the encoding detection will probably not be PEP263 but maybe BOM. PEP263 is an old scheme that does not really scale well to non-ASCII compatible encodings. On the other hand, CPython already recognizes UTF-8 BOM so it is natural to extend this to other Unicode encodings as well. Another possibility here is how CPython 3.8 detects [widechar Unicode in JSON](https://bugs.python.org/issue17909) (building on [RFC 4627](https://www.ietf.org/rfc/rfc4627.txt)). The assumptions there that the first character is ASCII or a quote are still true for a valid Python program. So in the end there is no need to extend support for PEP263 to widechar Unicode, especially that we already require BOM be present for widechar Unicode detection so PEP263 does not add anything useful.

Which approach do you prefer?

Reply:
If CPython ever does add support then who knows what detection mechanism they'll pick, so I would just go with whatever implementation is simpler (presumably this is checking the BOM?).

Reply:
Yes, checking the BOM is simpler. I will add a note to the "Differences with CPython" document stating that IronPython does support widechar Unicode based on BOM and that PEP263 does not apply in such case.